lib/micromachine.rb

```diff
@@ -44,9 +49,21 @@ def states
 private
 
   def change(event)
-    @state = transitions_for[event][@state]
+    previous_state, @state = @state, transitions_for[event][@state]
+
+    all_arguments = [event, previous_state, state]
```

Comment:
Hmmm. I tend to be very skeptical of claims of SOLID in the way that I’m also skeptical of claims of POLS for Ruby.

I can buy your argument for a `notify` method that must be triggered by `change`…and that means you have to basically call `notify(event, previous_state, state)`).

Your argument for the `state` reader means that the assignment at the top should be `previous_state, @state = state, transitions_for[event][state]`. I’m not sure there’s much increase in readability or reliability because `state` is *just* an `attr_reader`. IMO, direct access to an instance variable is exactly what a `private` method like `change` should be allowed and expected to do.

It doesn’t much matter to me, in the end, but there’s an inconsistency in use here.